yourbench/pipeline/answer_generation.py

```diff
@@ -0,0 +1,276 @@
+# yourbench/pipeline/answer_generation.py
+
+"""
+answer_generation.py
+
+Stage: answer_generation
+------------------------
+Generates answers to both single-shot and multi-hop questions for one or more 'scenarios'.
+Each scenario is appended to the *same* output subsets:
+  - single_shot_questions_with_answers
+  - multi_hop_questions_with_answers
+
+We distinguish the scenario by the 'answer_fashion' column in the final dataset (e.g. 'zero_shot', 'with_correct_chunk').
+
+Configuration Example (YAML):
+  answer_generation:
+    run: true
+    scenarios:
+      - zero_shot
+      - with_correct_chunk
+"""
+
+from typing import Any, Dict, List
+
+from loguru import logger
+
+from datasets import Dataset
+from yourbench.utils.prompts import GOLD_QA_USER_PROMPT, ZEROSHOT_QA_USER_PROMPT
+from yourbench.utils.dataset_engine import custom_load_dataset, custom_save_dataset
+from yourbench.utils.parsing_engine import extract_content_from_xml_tags
+from yourbench.utils.inference_engine import InferenceCall, run_inference
+
+
+def run(config: Dict[str, Any]) -> None:
+    """
+    Main entry point for the answer_generation stage.
+
+    Looks at config["pipeline"]["answer_generation"]["scenarios"] to see which answer-generation
+    modes to run. Currently supported: "zero_shot" and "with_correct_chunk".
+
+    Always concatenates new rows into:
+      - single_shot_questions_with_answers
+      - multi_hop_questions_with_answers
+
+    A column 'answer_fashion' indicates which scenario was used to generate the answer.
+    """
+    stage_cfg = config.get("pipeline", {}).get("answer_generation", {})
+    if not stage_cfg.get("run", False):
+        logger.info("answer_generation stage is disabled. Skipping.")
+        return
+
+    # Force dataset concatenation ON for this stage
+    logger.info("Forcing 'concat_if_exist = True' so new rows get appended.")
+    config.setdefault("hf_configuration", {})
+    config["hf_configuration"]["concat_if_exist"] = True
+
+    # Determine which scenarios to run
+    scenarios = stage_cfg.get("scenarios", ["zero_shot"])
+    logger.info(f"answer_generation scenarios = {scenarios}")
+
+    for scenario in scenarios:
+        logger.info(f"--- Starting scenario: {scenario} ---")
+
+        # Single-shot
+        _generate_answers_for_questions(
+            config=config,
+            source_subset="single_shot_questions",
+            scenario=scenario,
+            output_subset="single_shot_questions_with_answers",
+        )
+
+        # Multi-hop
+        _generate_answers_for_questions(
+            config=config,
+            source_subset="multi_hop_questions",
+            scenario=scenario,
+            output_subset="multi_hop_questions_with_answers",
+        )
```

Comment:
Why do you select both here? It should depend on the config and whether we generated multihop questions & single shot ones, no?

Reply:
updated the code so that single-shot answers are only generated if `single_shot_question_generation.run` is true, and similarly for multi-hop if `multi_hop_question_generation.run` is true. This way, we only generate answers for the question types actually produced.

of course, need to add further validation later, for special cases (perhaps multiple config files, later runs, etc), but this should do as an officially supported functionality